RavenDB-18800 Corax optimizations

ayende opened the pull request:
### Issue link

https://issues.hibernatingrhinos.com/issue/RavenDB-18959
https://issues.hibernatingrhinos.com/issue/RavenDB-18955
https://issues.hibernatingrhinos.com/issue/RavenDB-15895
https://issues.hibernatingrhinos.com/issue/RavenDB-18800
https://issues.hibernatingrhinos.com/issue/RavenDB-18987  
https://issues.hibernatingrhinos.com/issue/RavenDB-18837
https://issues.hibernatingrhinos.com/issue/RavenDB-19053
https://issues.hibernatingrhinos.com/issue/RavenDB-19052
https://issues.hibernatingrhinos.com/issue/RavenDB-19051


### Additional description

Fixing various performance issues with Corax, we are now looking a _significant_ performance different of Corax from Lucene.

![image](https://user-images.githubusercontent.com/116915/179317598-444dc537-662a-44c2-bf7b-651602384e6e.png)

maciejaszyk committed:
RavenDB-18959 Do not increase NumberOfEntries in metadataTree every single time

maciejaszyk committed:
RavenDB-18959  We don't have to assign ID from Bindinng because it's already there.

maciejaszyk committed:
RavenDB-18959 Create cached buffer for UTF8 conversion in IndexWriter for Analyzers

ayende committed:
RavenDB-18955 Corax: Degradation of indexing performance over time

Limit the number of records that we are allowed to scan when trying to find a better dictionary to ensure indexing liveliness and not slow things down when we have really big indexes.

ayende committed:
RavenDB-15895 On large database, index stats are not shown

When we need to estimate the indexing progress on very large indexes, we'll scan only about ~1 million entries to get an exact count.
Afterward, we'll use an estimated value, to reduce the cost of this estimation.

To give some context, if we have 750 million items, we have ~1.47 million pages just for the collection etags). That is ~11GB or so just for that.
We were scanning through all of that on a short duration, leading to both a lot of cost and not actually getting an answer ever.

ayende committed:
RavenDB-18959 - Turn a property into a field to save 0.33% of the total indexing cost

ayende committed:
RavenDB-18800 Investigate if we can drop bloom filters for Corax

ayende committed:
RavenDB-18959 - Reuse buffers while committing Corax entries

ayende committed:
RavenDB-18959 Ensure that we update IsAnalyzed when we update the analyzer for a field.

ayende committed:
RavenDB-18959  We don't need to sort the terms before inserting them. The same cost is already paid in the insert to the terms tree, and we *gain* performance by removing the sort

ayende committed:
RavenDB-18959  Don't set the default analyzer if the analyzer is already set

arekpalinski commented:
It looks that [SlowTests.MailingList.IndexTest.AvoidIndexWriterRecreation](https://jenkins.hibernatingrhinos.com/job/PR_Tests/2619/testReport/junit/SlowTests.MailingList/IndexTest/AvoidIndexWriterRecreation/) test is failing consistently.

maciejaszyk commented on src/Corax/IndexWriter.cs:
```diff
@@ -67,6 +68,9 @@ public class IndexWriter : IDisposable // single threaded, controlled by caller
         
 #pragma warning disable CS0169
         private Queue<long> _lastEntries; // keep last 256 items
+        private GrowablePooledBuffer<Slice> _sliceBuffer;
+        private GrowablePooledBuffer<long> _longBuffer;
+        private GrowablePooledBuffer<double> _doubleBuffer;
```
We don't use them anywhere

ayende replied:
correct, I removed them

ayende committed:
RavenDB-18959
- Reduce the frequency in which we are running defrags
- When merging pages, if they are using the same dictionary, avoid decoding & encoding them and copy them as-is

ayende commented:
Looking into the failure

ayende committed:
RavenDB-18959 Fixing AvoidIndexWriterRecreation test

ayende committed:
RavenDB-18959 Removing unused fields

ayende committed:
RavenDB-18987
- Refactoring to move local functions to instance functions (to make profiling easier)
- Change how we are managing deletes by moving to persistent set instead of an in memory one

redknightlois commented on src/Raven.Server/Documents/Indexes/Persistence/Corax/AnonymousCoraxDocumentConverter.cs:
```diff
@@ -45,21 +46,19 @@ public AnonymousCoraxDocumentConverter(Index index, bool storeValue = false)
             IndexField field;
-            if (knownFields.TryGetByFieldName(property.Key, out var binding))
+            try
             {
```
Code inside the `try...catch` is not subject to any type of optimization. Tread with care there (there is also prolog and epilog), if you can avoid it the codegen of the whole method would benefit.

ayende replied:
Done

redknightlois commented on src/Voron/Data/CompactTrees/CompactTree.cs:
```diff
@@ -549,19 +547,70 @@ private void MaybeMergeEntries(ref CursorState destinationState)
+            {
+                for (; sourceKeysCopied < sourceHeader->NumberOfEntries; sourceKeysCopied++)
+                {
+                    if (MoveEntryWithReEncoding(decodeBuffer, encodeBuffer, ref destinationState, entries) == false)
```
Do you have a profile that shows a material difference here? Because, given that we are executing for many (hundreds) of iterations every time, the branch detector will take care of it. If there is a difference, I will look into what it is triggering the JIT to underperform there, because there should be no differences.

ayende replied:
Done

redknightlois commented on src/Corax/Utils/GrowablePooledBuffer.cs:
```diff
@@ -0,0 +1,45 @@
+namespace Corax.Utils;
+
+public struct GrowablePooledBuffer<T> : IDisposable
+{
```
Should be removed if not used. And if we are going to keep it (because it is properly tested) it must go in `Sparrow.Memory`

ayende replied:
done

redknightlois commented:
Yes, I know... but I just came from the ski course and it is snowing like crap. Nothing better to do than check gmail and find a PR named "Optimizations" :D

ayende committed:
RavenDB-18987
- Optimize entries deletes in Corax
- For Small sets, we'll update the value in place (instead of rebuilding from scratch)
- Modify everything that work with small sets to account for the size != the number of items
- Entries count should be long, we now support > 2b records

ayende committed:
RavenDB-18987 Avoid overwriting the term set if it is the same as before

ayende committed:
RavenDB-18987 - PR Comments

arekpalinski commented on src/Sparrow.Server/ByteString.cs:
```diff
@@ -446,7 +446,7 @@ public ReadOnlySpan<byte> ToReadOnlySpan()
-        public Span<byte> ToSpan()
+        public readonly Span<byte> ToSpan()
         {
             return new Span<byte>(Ptr, Length);
```
@ayende can you please check the warnings here?

ayende replied:
yes, reverted.

arekpalinski commented:
As I see the approach in handling the deletions got changed. I believe we should bump the internal Corax version number since this probably requires existing indexes to be reset. @maciejaszyk can you point where is that number?

arekpalinski commented:
Last tests run seems to be crashed (https://jenkins.hibernatingrhinos.com/job/PR_Tests/2643/console):

```
02:55:18  This test may, or may not be the source of the crash.
02:55:19  [Pipeline] echo
02:55:19  [LOG] encountered hudson.AbortException: script returned exit code 1
```

I have scheduled a few more runs:
https://jenkins.hibernatingrhinos.com/job/PR_Tests/2646/
https://jenkins.hibernatingrhinos.com/job/PR_Tests/2647/
https://jenkins.hibernatingrhinos.com/job/PR_Tests/2648/

maciejaszyk commented:
> As I see the approach in handling the deletions got changed. I believe we should bump the internal Corax version number since this probably requires existing indexes to be reset. @maciejaszyk can you point where is that number?

`Corax.Constants.IndexWriter.Version`

ayende committed:
RavenDB-18837
- Changing Corax indexing version (backward compact is not maintained)
- Support storage report for Corax
- Now also tracking the total number of entries per Container

ayende committed:
RavenDB-18837 - Ensure that when we merge leaf pages in a set, we'll merge only pages from the same baseline.

arekpalinski commented:
1. [SlowTests.Issues.RavenDB_14576.JavascriptIndexShouldPassContextToTypeConverter](https://jenkins.hibernatingrhinos.com/job/PR_Tests/2681/testReport/SlowTests.Issues/RavenDB_14576/JavascriptIndexShouldPassContextToTypeConverter/) failed in 2 consecutive runs
2. https://issues.hibernatingrhinos.com/issue/RavenDB-19001/Corax-Indexing-errors-after-performance-optimizations

ayende committed:
RavenDB-18987 Indexing deletions and additions are now both using bulk mode to optimize costs.

ayende commented:
I updated the version number, looking into the test failures.

ayende committed:
RavenDB-18987 Reserve small set posting lists with 32 bytes of additional space (to amortize costs of adding items to the posting list). Reuse space when we need to remove data from the posting list.

ayende committed:
RavenDB-18987 Fixing sort order on insert / delete

ayende committed:
RavenDB-18987 - Fixing indexing batch optimizations

ayende committed:
* RavenDB-18987 Use ReusableTokenStream instead of TokenStream to *massively* reduce the number of allocations we have

ayende committed:
RavenDB-18880 Fixing estimation of number of entries for large trees.
We now let 10 seconds to go by before we do estimation. Assuming that we'll use this to slowly load things to memory, that gives us slowly more accurate value.

ayende committed:
RavenDB-18880 Ensuring that we wait for up to 10 seconds to evaluate how many documents we have remaining for specified work. Afterward, we'll just estimate the amount of work, to avoid stalling the user.

arekpalinski commented on src/Sparrow.Server/ByteString.cs:
```diff
@@ -446,7 +446,7 @@ public ReadOnlySpan<byte> ToReadOnlySpan()
 
         [MethodImpl(MethodImplOptions.AggressiveInlining)]
-        public Span<byte> ToSpan()
+        public readonly Span<byte> ToSpan()
```
It looks the fix for this warning was reverted with other changes

ayende replied:
Fixed

ayende committed:
RavenDB-18800 - Fixing compiler warning

ayende committed:
RavenDB-19051 Corax - avoid using QueryContext.Pool, use native memory instead

ayende committed:
RavenDB-19052 - Removing suspicious comparers

ayende committed:
RavenDB-19041 - Attempt to reduce decompression costs by avoiding span / fixed, etc.

ayende committed:
RavenDB-18800
 - Reduce number of times we index into the journal page info
 - Avoid (expensive) memmove call in favor of doing copies manually (expected to be very small)

ayende committed:
RavenDB-19053 Remove Unsafe.AsRef<int>(Unsafe.AsPointer()) usage

arekpalinski commented on src/Raven.Server/Utils/MetricCacher.cs:
```diff
@@ -67,86 +68,55 @@ public T GetValue<T>(string key, Func<T> factory = null)
-                first = _lastRefresh == DateTime.MinValue;
 
-                return first || SystemTime.UtcNow - _lastRefresh >= _refreshRate;
+                return _value;
```
You might return the stale value here because the just started task could not update it yet. I'm not sure if that matters much but this is a change in comparison to the current behavior.

ayende replied:
Yes, that is by design, because we are currently still processing it.
We are okay in returning a stale value, I believe.

ppekrol commented on src/Raven.Server/Utils/MetricCacher.cs:
```diff
@@ -67,86 +68,55 @@ public T GetValue<T>(string key, Func<T> factory = null)
+                    return _value; // return current value, while it is being computed
 
-            public object Value { get; private set; }
+                if (currentTask.IsFaulted)
```
this is problematic because inability to start a task (e.g. low memory) or any error, will cause infinite exceptions here, so basically you will force users to restart the server

ayende replied:
You are correct, fixed.

arekpalinski commented on src/Raven.Server/Documents/CollectionRunner.cs:
```diff
@@ -190,11 +190,11 @@ protected long GetTotalCountForCollection(DocumentsOperationContext context, str
             }
 
-            Database.DocumentsStorage.GetNumberOfDocumentsToProcess(context, collectionName, 0, out long totalCount);
+            Database.DocumentsStorage.GetNumberOfDocumentsToProcess(context, collectionName, 0, out long totalCount, Stopwatch.StartNew());
```
Why do we need to pass `Stopwatch.StartNew()`? It looks a bit suspicious

ayende replied:
We need to ensure that we aren't spending too long here.
That measures that total time that we spend waiting to run this computation

arekpalinski commented on src/Voron/Data/Fixed/FixedSizeTree.cs:
```diff
@@ -1653,7 +1653,7 @@ public override string ToString()
         }
 
-        public long GetNumberOfEntriesAfter(TVal value, out long totalCount)
+        public long GetNumberOfEntriesAfter(TVal value, out long totalCount, Stopwatch overallDuration)
```
My concern is that now this can return inconsistent results since the method signature doesn't indicate that it can be estimated. I'm worried about someone using it for different purpose in the future and it will work fine during development but not in production.

ayende replied:
This _does_ return wrong results, but the nice thing here is that it fixed itself, since each time it loads more of the relevant pieces to memory.

Do you have a good idea on how we can handle that? I can make it return an object that says `Estimated`, of course. But not sure how useful that would be.

ayende committed:
RavenDB-19053 - Fixing how we handle decoding of values without Unsafe.AsPointer. Making sure we always abort on the end of the term
